core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -141,25 +137,15 @@ private static void putReportedMetrics(IndexedSet<Metric> metricSet, String inst
       // If a metric is COUNTER, the value sent via RPC should be the incremental value; i.e.
       // the amount the value has changed since the last RPC. The master should equivalently
       // increment its value based on the received metric rather than replacing it.
-      if (metric.getMetricType() == MetricType.COUNTER) {
-        // FULL_NAME_INDEX is a unique index, so getFirstByField will return the same results as
-        // getByField
-        Metric oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        double oldVal = oldMetric == null ? 0.0 : oldMetric.getValue();
-        Metric newMetric = new Metric(metric.getInstanceType(), metric.getHostname(),
-            metric.getMetricType(), metric.getName(), oldVal + metric.getValue());
-        for (Map.Entry<String, String> tag : metric.getTags().entrySet()) {
-          newMetric.addTag(tag.getKey(), tag.getValue());
-        }
-        metricSet.removeByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        newMetrics.add(newMetric);
+      Metric oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
```

Comment:
instead of reading first, then adding, can we just call `add` directly? The result of `add` will be `true` if it successfully add the new item, and `false` if it already exists. Then, if it didn't exist, we can "update" the metric value itself but setting it or incrementing it.

I think if we make these changes, we can even remove the `synchronized` block from the put methods above.